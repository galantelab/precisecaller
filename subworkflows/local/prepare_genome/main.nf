//
// Prepare reference genome files
//

include { BWA_INDEX                      } from '../../../modules/nf-core/bwa/index/main'
include { GATK4_CREATESEQUENCEDICTIONARY } from '../../../modules/nf-core/gatk4/createsequencedictionary/main'
include { GUNZIP as GUNZIP_FASTA         } from '../../../modules/nf-core/gunzip'
include { SAMTOOLS_FAIDX                 } from '../../../modules/nf-core/samtools/faidx/main'

workflow PREPARE_GENOME {
    take:
    fasta      // file: [mandatory] /path/to/genome.fasta
    fasta_fai  // file: [optional]  /path/to/genome.fasta.fai
    dict       // file: [optional]  /path/to/genome.dict
    bwa        // file: [optional]  /path/to/bwa/

    main:
    ch_versions = Channel.empty()
    fasta_file  = file(fasta, checkIfExists: true)

    ch_fasta = Channel.empty()
    if (fasta.endsWith('.gz')) {
        GUNZIP_FASTA([[id:"${fasta_file.baseName}"], fasta_file])
        ch_fasta     = GUNZIP_FASTA.out.gunzip
        ch_versions  = ch_versions.mix(GUNZIP_FASTA.out.versions)
    } else {
        ch_fasta = Channel.value([[id:"${fasta_file.baseName}"], fasta_file])
    }

    ch_fasta_fai = Channel.empty()
    if (fasta_fai) {
        ch_fasta_fai = Channel.value([[id:"fai"], file(fasta_fai, checkIfExists: true)])
    } else {
        SAMTOOLS_FAIDX(ch_fasta, [[ id:'no_fai' ], []], false)
        ch_fasta_fai = SAMTOOLS_FAIDX.out.fai
        ch_versions  = ch_versions.mix(SAMTOOLS_FAIDX.out.versions)
    }

    ch_dict = Channel.empty()
    if (dict) {
        ch_dict = Channel.value([[id:"dict"], file(dict, checkIfExists: true)])
    } else {
        GATK4_CREATESEQUENCEDICTIONARY(ch_fasta)
        ch_dict     = GATK4_CREATESEQUENCEDICTIONARY.out.dict
        ch_versions = ch_versions.mix(GATK4_CREATESEQUENCEDICTIONARY.out.versions)
    }

    ch_bwa = Channel.empty()
    if (bwa) {
        ch_bwa = Channel.value([[id:"bwa"], file(bwa, checkIfExists: true)])
    } else {
        BWA_INDEX(ch_fasta)
        ch_bwa       = BWA_INDEX.out.index
        ch_versions  = ch_versions.mix(BWA_INDEX.out.versions)
    }

    emit:
    fasta     = ch_fasta      // channel: [ val(meta), [ fasta ] ]
    fasta_fai = ch_fasta_fai  // channel: [ val(meta), [ fai   ] ]
    dict      = ch_dict       // channel: [ val(meta), [ dict  ] ]
    bwa       = ch_bwa        // channel: [ val(meta), [ bwa   ] ]
    versions  = ch_versions   // channel: [ versions.yml ]
}
