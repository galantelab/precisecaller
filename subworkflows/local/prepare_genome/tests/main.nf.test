// nf-core subworkflows test prepare_genome
nextflow_workflow {
    name "Test Subworkflow PREPARE_GENOME"
    script "../main.nf"
    workflow "PREPARE_GENOME"

    tag "subworkflows"
    tag "subworkflows_"
    tag "subworkflows/fasta_prepare_references"

    test("fasta - prepare fasta reference") {
        config "./nextflow.config"

        when {

            workflow {
                """
                input[0] = params.local_testdata_base_path + 'references/Homo_sapiens/Gencode/hg38_chr22/Sequence/WholeGenomeFasta/genome.fa'
                input[1] = null
                input[2] = null
                input[3] = null
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }
}
