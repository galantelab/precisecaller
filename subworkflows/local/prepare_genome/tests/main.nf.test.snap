{
    "fasta.gz - prepare fasta reference": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "genome.fa"
                        },
                        "genome.fa:md5,96abec2e5bbf15e88af5c6adbb6fb33e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "genome.fa"
                        },
                        "genome.fa.fai:md5,889ed0b342d3854343aebb08cd0a9c6d"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "genome.fa"
                        },
                        "genome.dict:md5,41f4d1eb30aa93e078a2997d9768e93f"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "genome.fa"
                        },
                        [
                            "genome.amb:md5,4a9a52d98d38ddbaa438b71c68d24446",
                            "genome.ann:md5,b4aaa436f9ea7fdf9dc36d8914912f6c",
                            "genome.bwt:md5,fcc50591c1320abc5f87b6087aed93e7",
                            "genome.pac:md5,2f93530bfcb6be9dd4918b52249e5db2",
                            "genome.sa:md5,a8c6c3224d4e167a3329eb69f065aed9"
                        ]
                    ]
                ],
                "4": [
                    "versions.yml:md5,265795dcfabc74e982198a30abb773d5",
                    "versions.yml:md5,6500dc776f23be2846f35446e2e38897",
                    "versions.yml:md5,7303f194674b485c6914e28422822a12",
                    "versions.yml:md5,e2315fc397ebc95e618a8f7fcfb8dc12"
                ],
                "bwa": [
                    [
                        {
                            "id": "genome.fa"
                        },
                        [
                            "genome.amb:md5,4a9a52d98d38ddbaa438b71c68d24446",
                            "genome.ann:md5,b4aaa436f9ea7fdf9dc36d8914912f6c",
                            "genome.bwt:md5,fcc50591c1320abc5f87b6087aed93e7",
                            "genome.pac:md5,2f93530bfcb6be9dd4918b52249e5db2",
                            "genome.sa:md5,a8c6c3224d4e167a3329eb69f065aed9"
                        ]
                    ]
                ],
                "dict": [
                    [
                        {
                            "id": "genome.fa"
                        },
                        "genome.dict:md5,41f4d1eb30aa93e078a2997d9768e93f"
                    ]
                ],
                "fasta": [
                    [
                        {
                            "id": "genome.fa"
                        },
                        "genome.fa:md5,96abec2e5bbf15e88af5c6adbb6fb33e"
                    ]
                ],
                "fasta_fai": [
                    [
                        {
                            "id": "genome.fa"
                        },
                        "genome.fa.fai:md5,889ed0b342d3854343aebb08cd0a9c6d"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,265795dcfabc74e982198a30abb773d5",
                    "versions.yml:md5,6500dc776f23be2846f35446e2e38897",
                    "versions.yml:md5,7303f194674b485c6914e28422822a12",
                    "versions.yml:md5,e2315fc397ebc95e618a8f7fcfb8dc12"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-13T17:47:05.820589"
    },
    "fasta - prepare fasta reference": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "genome"
                        },
                        "/precisecaller/test-datasets/references/Homo_sapiens/Gencode/hg38_chr22q12.1/Sequence/WholeGenomeFasta/genome.fa"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "genome"
                        },
                        "genome.fa.fai:md5,889ed0b342d3854343aebb08cd0a9c6d"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "genome"
                        },
                        "genome.dict:md5,41f4d1eb30aa93e078a2997d9768e93f"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "genome"
                        },
                        [
                            "genome.amb:md5,4a9a52d98d38ddbaa438b71c68d24446",
                            "genome.ann:md5,b4aaa436f9ea7fdf9dc36d8914912f6c",
                            "genome.bwt:md5,fcc50591c1320abc5f87b6087aed93e7",
                            "genome.pac:md5,2f93530bfcb6be9dd4918b52249e5db2",
                            "genome.sa:md5,a8c6c3224d4e167a3329eb69f065aed9"
                        ]
                    ]
                ],
                "4": [
                    "versions.yml:md5,265795dcfabc74e982198a30abb773d5",
                    "versions.yml:md5,6500dc776f23be2846f35446e2e38897",
                    "versions.yml:md5,7303f194674b485c6914e28422822a12"
                ],
                "bwa": [
                    [
                        {
                            "id": "genome"
                        },
                        [
                            "genome.amb:md5,4a9a52d98d38ddbaa438b71c68d24446",
                            "genome.ann:md5,b4aaa436f9ea7fdf9dc36d8914912f6c",
                            "genome.bwt:md5,fcc50591c1320abc5f87b6087aed93e7",
                            "genome.pac:md5,2f93530bfcb6be9dd4918b52249e5db2",
                            "genome.sa:md5,a8c6c3224d4e167a3329eb69f065aed9"
                        ]
                    ]
                ],
                "dict": [
                    [
                        {
                            "id": "genome"
                        },
                        "genome.dict:md5,41f4d1eb30aa93e078a2997d9768e93f"
                    ]
                ],
                "fasta": [
                    [
                        {
                            "id": "genome"
                        },
                        "/precisecaller/test-datasets/references/Homo_sapiens/Gencode/hg38_chr22q12.1/Sequence/WholeGenomeFasta/genome.fa"
                    ]
                ],
                "fasta_fai": [
                    [
                        {
                            "id": "genome"
                        },
                        "genome.fa.fai:md5,889ed0b342d3854343aebb08cd0a9c6d"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,265795dcfabc74e982198a30abb773d5",
                    "versions.yml:md5,6500dc776f23be2846f35446e2e38897",
                    "versions.yml:md5,7303f194674b485c6914e28422822a12"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-13T17:45:58.268773"
    }
}